File: python-sim/tests/test_crc16.py
from hftext.codec import encode_text_to_symbols, pack_symbols_to_bytes
from hftext.crc16 import crc16_ccitt_false


def test_crc16_ccitt_false_known_vector():
    assert crc16_ccitt_false(b"123456789") == 0x29B1


def test_crc16_empty_payload_uses_initial_value():
    assert crc16_ccitt_false(b"") == 0xFFFF


def test_crc16_changes_when_data_changes():
    original = crc16_ccitt_false(b"hftext")
    changed = crc16_ccitt_false(b"hftexu")

    assert original != changed


def test_crc16_can_run_over_packed_payload_symbols():
    symbols = encode_text_to_symbols("pu5lrk Teste")
    payload_bytes = pack_symbols_to_bytes(symbols)

    assert crc16_ccitt_false(payload_bytes) == crc16_ccitt_false(payload_bytes)
    assert crc16_ccitt_false(payload_bytes) != crc16_ccitt_false(payload_bytes + b"\x00")

File: core/tests/test_crc16.cpp
#include "hftext_crc16.h"

#include <cassert>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

std::vector<std::uint8_t> bytesFromString(const std::string& text) {
    return {text.begin(), text.end()};
}

}  // namespace

int main() {
    const auto known = bytesFromString("123456789");
    assert(hftext::crc16CcittFalse(known) == 0x29B1);

    const std::vector<std::uint8_t> empty;
    assert(hftext::crc16CcittFalse(empty) == 0xFFFF);

    const auto original = bytesFromString("hftext");
    const auto changed = bytesFromString("hftexu");
    assert(hftext::crc16CcittFalse(original) != hftext::crc16CcittFalse(changed));

    const std::vector<std::uint8_t> payloadBytes = {0x04, 0x20, 0xC0};
    assert(hftext::crc16CcittFalse(payloadBytes) == hftext::crc16CcittFalse(payloadBytes));

    auto modifiedPayload = payloadBytes;
    modifiedPayload.push_back(0);
    assert(hftext::crc16CcittFalse(payloadBytes) != hftext::crc16CcittFalse(modifiedPayload));

    bool nullRejected = false;
    try {
        (void)hftext::crc16CcittFalse(nullptr, 1);
    } catch (const std::invalid_argument&) {
        nullRejected = true;
    }
    assert(nullRejected);
}
